### PR TITLE
AJ-939, IA-4193 source workspace id parameter

### PR DESCRIPTION
**Pull request by calypsomatic**
Note: changes to `Dockerfile` and `reference.conf` are temporary for testing and will not be a part of the final PR.
Changes for [IA-4193](https://broadworkbench.atlassian.net/browse/IA-4193) in support of [AJ-939](https://broadworkbench.atlassian.net/browse/AJ-939): WDS needs to receive a `sourceWorkspaceId`. This adds the `sourceWorkspaceId` as a parameter to the `createApp` API, and then adds it to the helm chart overrides (if present).

[IA-4193]: https://broadworkbench.atlassian.net/browse/IA-4193?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[AJ-939]: https://broadworkbench.atlassian.net/browse/AJ-939?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: http/src/main/resources/swagger/api-docs.yaml [hunk 1]
```diff
@@ -3858,6 +3858,9 @@ components:
           items:
             type: string
           description: Extra arguments to pass to the application. Only used if appType is CUSTOM.
+        sourceWorkspaceId:
+          type: string
+          description: Id of source workspace if app is a clone.
```
Comment (LizBaldo) [inline http/src/main/resources/swagger/api-docs.yaml]: Maybe say that this is optional

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AKSInterpreter.scala [hunk 2]
```diff
@@ -116,7 +116,7 @@ class AKSInterpreter[F[_]](config: AKSInterpreterConfig,
       )
 
       _ <- logger.info(ctx.loggingCtx)(
-        s"Begin app creation for app ${params.appName.value} in cloud context ${params.cloudContext.asString}"
+        s"Begin app creation for app ${params.appName.value} in cloud context ${params.cloudContext.asString} with variables ${app.customEnvironmentVariables}"
```
Comment (LizBaldo) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AKSInterpreter.scala]: Looks like a leftover from the previous PR, but it really can't hurt to add this to the logs 👍

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AKSInterpreter.scala [hunk 3]
```diff
@@ -510,8 +512,14 @@ class AKSInterpreter[F[_]](config: AKSInterpreterConfig,
         // general configs
         raw"fullnameOverride=$appChartPrefix-${release.asString}",
         raw"instrumentationEnabled=${config.coaAppConfig.instrumentationEnabled}"
-      ).mkString(",")
-    )
+      )
+
+    val updatedLs = sourceWorkspaceId match { // TODO move to buildWdsChartOverrideValues after WDS chart migration
+      case Some(value) => ls ::: List(raw"config.sourceWorkspaceId=${value.value}")
```
Comment (LizBaldo) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AKSInterpreter.scala]: That makes sense to me, mad at scala for having list concatenation defined as ::: though 😡

**Review by LizBaldo — COMMENTED** [review]
LGTM! Left very minor comments, ping me when you want a final 👍

**Comment by LizBaldo** [comment]
@davidangb do you need me to review something specific again? I did not give a 👍 yet given that the PR is a draft.

**Comment by davidangb** [comment]
@LizBaldo whoops sorry! That was a misclick, please ignore that.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AKSInterpreter.scala [hunk 4]
```diff
@@ -466,9 +468,10 @@ class AKSInterpreter[F[_]](config: AKSInterpreterConfig,
                                                      storageContainer: StorageContainerResponse,
                                                      batchAccountKey: BatchAccountKey,
                                                      applicationInsightsConnectionString: String,
-                                                     appChartPrefix: String
-  ): Values =
-    Values(
+                                                     appChartPrefix: String,
+                                                     sourceWorkspaceId: Option[WorkspaceId]
+  ): Values = {
+    val ls =
```
Comment (aaronkanzer) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AKSInterpreter.scala]: Could we use a more descriptive variable name than `ls`?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AKSInterpreter.scala [hunk 5]
```diff
@@ -510,8 +513,14 @@ class AKSInterpreter[F[_]](config: AKSInterpreterConfig,
         // general configs
         raw"fullnameOverride=$appChartPrefix-${release.asString}",
         raw"instrumentationEnabled=${config.coaAppConfig.instrumentationEnabled}"
-      ).mkString(",")
-    )
+      )
+
+    val updatedLs = sourceWorkspaceId match { // TODO remove after WDS chart migration
+      case Some(value) => ls ::: List(raw"provenance.sourceWorkspaceId=${value.value}")
```
Comment (aaronkanzer) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AKSInterpreter.scala]: Trying to track this down, but where does `provenance` live in our Helm chart values exactly?

Comment (aaronkanzer) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AKSInterpreter.scala]: ah jk -- found it in one of your other PRs... 🤦

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AKSInterpreter.scala (hunk 4 above)
Comment (calypsomatic) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AKSInterpreter.scala]: Sure why not

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/AppComponent.scala [hunk 6]
```diff
@@ -204,7 +207,8 @@ object appQuery extends TableQuery(new AppTable(_)) {
         diskOpt.map(_.id),
         if (saveApp.app.customEnvironmentVariables.isEmpty) None else Some(saveApp.app.customEnvironmentVariables),
         saveApp.app.descriptorPath,
-        if (saveApp.app.extraArgs.isEmpty) None else Some(saveApp.app.extraArgs)
+        if (saveApp.app.extraArgs.isEmpty) None else Some(saveApp.app.extraArgs),
+        if (saveApp.app.sourceWorkspaceId.isEmpty) None else saveApp.app.sourceWorkspaceId
```
Comment (davidangb) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/AppComponent.scala]: I think this is redundant. If `saveApp.app.sourceWorkspaceId.isEmpty` is true, then `saveApp.app.sourceWorkspaceId` will be a `None`. So you can just use:
```suggestion
        saveApp.app.sourceWorkspaceId
```
similar to how `saveApp.app.descriptorPath` is specified at line 209.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AKSInterpreter.scala [hunk 7]
```diff
@@ -514,8 +517,14 @@ class AKSInterpreter[F[_]](config: AKSInterpreterConfig,
...
         raw"fullnameOverride=$appChartPrefix-${release.asString}",
         raw"instrumentationEnabled=${config.coaAppConfig.instrumentationEnabled}"
-      ).mkString(",")
-    )
+      )
+
+    val updatedLs = sourceWorkspaceId match { // TODO remove after WDS chart migration
+      case Some(value) => valuesList ::: List(raw"provenance.sourceWorkspaceId=${value.value}")
+      case None        => valuesList
+    }
```
Comment (davidangb) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AKSInterpreter.scala:525]: This is perfectly fine as-is, no need to change anything. For the sake of discussing Scala goodness, some thoughts:

You could rewrite this using `.map`, which is a bit more functional-programming-esque than an explicit `match` statement here:
```scala
val updatedLs = valuesList ++ sourceWorkspaceId.map(v => raw"provenance.sourceWorkspaceId=${v.value}").toList
```
This also relies on the behavior of `Option.toList` to create an empty list for None and a single-value list for Some.

~Also, beware the difference between `:::` and `++` when concatenating lists. The former prepends the right-hand value to the list, while the latter appends. In this particular case I don't think it matters at all, but in other cases it certainly could, and might lead to confusion. At lists of this size the performance impact should be nonexistent.~

EDIT: this last paragraph, now stricken, is incorrect … please ignore!

**Review by davidangb — COMMENTED** [review]
minor comments inline, looking good!

I don't have a great sense in Leo if there's anything you missed - the code changes here look good.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AKSInterpreter.scala (hunk 7 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AKSInterpreter.scala:525]: +1 David's suggestion is a bit more idiomatic, but also ok as is.

Does the helm chart handle no value for `provenance.sourceWorkspaceId` if the app is not a clone? E.g. it doesn't need some default value?

**Review by rtitle — APPROVED** [review]
LGTM! Thanks for making the DB migration too

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AKSInterpreter.scala (hunk 7 above)
Comment (calypsomatic) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AKSInterpreter.scala:525]: In my testing, it was injecting the placeholder value - i.e. "RUNTIME_PARAMETER", so I used a blank string "".  I'm happy to hear a better solution.

**Review by LizBaldo — APPROVED** [review]
LGTM! :shipit:

**Review by davidangb — APPROVED** [review]
lgtm, nice job!

**Pull request merged**